api/proquest/client.py

```diff
@@ -0,0 +1,667 @@
+        :type hits_per_page: int
+        """
+        self._logger.info(
+            "Started downloading page # {0} ({1} hits) of a paginated OPDS 2.0 feed from {2}".format(
+                page, hits_per_page, configuration.books_catalog_service_url
+            )
+        )
+
+        parameters = {"page": page, "hitsPerPage": hits_per_page}
+        response, response_json = self._send_request(
+            configuration,
+            "get",
+            configuration.books_catalog_service_url,
+            parameters,
+            response_must_be_json=True,
+        )
+
+        self._logger.info(
+            "Finished downloading page # {0} ({1} hits) of a paginated OPDS 2.0 feed from {2}".format(
+                page, hits_per_page, configuration.books_catalog_service_url
+            )
+        )
+
+        if self.RESPONSE_OPDS_FEED_FIELD not in response_json:
+            raise ProQuestAPIMissingJSONPropertyError(
+                response, self.RESPONSE_OPDS_FEED_FIELD
+            )
+
+        return response_json[self.RESPONSE_OPDS_FEED_FIELD]
+
+    def download_feed_page(self, db, page, hits_per_page):
+        """Download a single page of a paginated OPDS 2.0 feed.
+
+        :param db: Database session
+        :type db: sqlalchemy.orm.session.Session
+
+        :param page: Page index (max = 32,766)
+        :type page: int
+
+        :param hits_per_page: Number of publications on a single page (max = 32,766)
+        :type hits_per_page: int
+        """
+        if not is_session(db):
+            raise ValueError('"db" argument must be a valid SQLAlchemy session')
+        if not isinstance(page, int):
+            raise ValueError('"page" argument must be an integer')
+        if page < 0 or page > self.MAX_PAGE_INDEX:
+            raise ValueError(
+                "Page argument must a non-negative number less than {0}".format(
+                    self.MAX_PAGE_INDEX
+                )
+            )
+        if not isinstance(hits_per_page, int):
+            raise ValueError('"hits_per_page" argument must be an integer')
+        if hits_per_page < 0 or hits_per_page > self.MAX_PAGE_SIZE:
+            raise ValueError(
+                "Hits per page argument must a non-negative number less than {0}".format(
+                    self.MAX_PAGE_SIZE
+                )
+            )
+
+        self._logger.info(
+            "Started downloading page # {0} ({1} hits) of a paginated OPDS 2.0 feed ".format(
+                page, hits_per_page
+            )
+        )
+
+        with self._get_configuration(db) as configuration:
+            feed = self._download_feed_page(configuration, page, hits_per_page)
+
+            self._logger.info(
+                "Finished downloading page # {0} ({1} hits) of a paginated OPDS 2.0 feed".format(
+                    page, hits_per_page
+                )
+            )
+
+            return feed
+
+    def download_all_feed_pages(self, db):
```

Comment:
The problem is that ProQuest's OPDS feed doesn't contain `next` links and we have to iterate through it using the API.
Also, unfortunately, we have to download the whole feed all the time because only by doing this we will be able to handle removals. There is no logic for it yet, I'll be adding it a bit later. It will basically contain the following steps we discussed before:
1. Create a set S1 of all current ProQuest IDs
2. Walk through the whole feed and save all the IDs from the feed in a separate set S2
3. Subtract S2 from S1, the resulting difference will contain IDs which needs to be removed

Reply:
Even though the feed doesn't contain `next` links, you know how to construct the "link that should be processed next". I think you could do that work inside `next_links`, instead of making `next_links` a no-op and doing the work elsewhere.

Having to download the whole feed every time is unfortunate. How big is this feed likely to be? My concern is that a network error at any point in the process will invalidate the entire operation, so it may never actually complete.

Reply:
Unfortunately, it's not enough to just construct URL and pass it as a `next_link`. The ProQuest's API returns a JSON document containing an OPDS 2.0 feed as a field inside this document. `ProQuestAPIClient` contains special logic parsing API's responses and validating them. That's why I would like to keep the feed downloading logic there instead of extending `OPDSImporter`.
The feed can be quite large, at the moment it's around 38 Mb.

Reply:
My advice based on experience with other APIs is it's better to download a single document, even if it takes a really long time, than to try to download 100 smaller documents.